### PR TITLE
Add tighter bound to VertexAI retry filter

### PR DESCRIPTION
**Pull request by jrmccluskey**
Adds and tests a utility function that retries the request loop only on HTTP 4xx errors. Previous behavior would trigger a retry on any exception being raised. Also adds an explicit callout for when an unexpected exception is raised at an error level for clarity.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by jrmccluskey** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/ml/inference/vertex_ai_inference.py [hunk 1]
```diff
@@ -41,6 +41,22 @@
 # pylint: disable=line-too-long
 
 
+def retry_on_gcp_client_error(exception):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/ml/inference/vertex_ai_inference.py:44]: nit: prepend with _ as this is an internal helper. Alternatively, could also inline this function via a `lambda` & include the rationale in the PR (no strong opinion here).

**Review by tvalentyn — APPROVED** [review]

**Comment by jrmccluskey** [comment]
The linting issue is a known problem elsewhere in the code base being fixed in #27451, merging once other precommits pass

File: sdks/python/apache_beam/ml/inference/vertex_ai_inference.py (hunk 1 above)
Comment (jrmccluskey) [reply sdks/python/apache_beam/ml/inference/vertex_ai_inference.py]: Done

**Pull request merged**